### PR TITLE
Add proposed ways to write bound functions within classes

### PR DESCRIPTION
**Pull request by jiangyue-eb**
@kaylieEB @BenAtEventbrite

### CODE REVIEW

File: es6/README.md [hunk 1]
```diff
@@ -704,6 +704,69 @@ Button.propTypes = {
...
+    }
+
+    render() {
+        return (
+            <button onClick={this._handleOnClick}/>
+        );
+    }
+}
+
+// bad: use `::func` to bind events need to be removed later
```
Comment (benmvp) [inline es6/README.md]: I wouldn't even bother w/ this example because only super old code uses it

File: es6/README.md [hunk 2]
```diff
@@ -704,6 +704,69 @@ Button.propTypes = {
 
 Both declarative property syntaxes are not a part of the ES2015 specification and are a in the midst of the ECMAScript proposal approval process. Currently they are sitting in Stage 3. For details, check out [ECMAScript Class Fields and Static Properties](https://github.com/jeffmo/es-class-fields-and-static-properties).
 
+For bound methods within classes, we recommend to use fat arrow syntax to write functions need current context:
```
Comment (benmvp) [inline es6/README.md]: So then are you suggesting that we not do:

```js
class Button extends React.Component {
    _handleOnClick() {
        // handle button click
    }

    render() {
        return (
             <button onClick={this._handleOnClick.bind(this)} />
        );
    }
}
```

either?

If so we'll have to update all of the other examples in this doc to reflect this new pattern.

File: es6/README.md [hunk 3]
```diff
@@ -704,6 +704,69 @@ Button.propTypes = {
...
+
+```js
+// good
+class Button extends React.Component {
+    _handleOnClick = () => {
+        //handle button click.
+    }
+
+    render() {
```
Comment (benmvp) [inline es6/README.md]: Should `render` (and other lifecycle methods) also be bound too?

```
render = () => (
    <button onClick={this._handleOnClick} />
)
```

File: es6/README.md (hunk 2 above)
Comment (jiangyue-eb) [reply es6/README.md]: I think this is fine, my only intention is to replace the `::func` case because it might create some bugs where we want to refer to the same callback but `::func` basically means `func.bind` and it returns a new instance each time.

File: es6/README.md (hunk 1 above)
Comment (jiangyue-eb) [reply es6/README.md]: I saw there were some places in `seat_designer` repo, maybe relatively old repo? And it had a bug I found where the `removeEventHandler` didn't remove the resize handler when `componentWillUnmount` because of the usage of `::func`.

Comment (benmvp) [reply es6/README.md]: Yeah we definitely shouldn't be using the bind operator. That repo is old and didn't conform to the consistency we've been trying to put in place

File: es6/README.md (hunk 3 above)
Comment (jiangyue-eb) [reply es6/README.md]: I like only using this functions if the function needs to be bound, e.g. event handlers, since this syntax is more concise and reduce the risk that some issues raised by `.bind`.

File: es6/README.md (hunk 2 above)
Comment (benmvp) [reply es6/README.md]: If that's the case then, I wouldn't bother with addition because the two "bad" examples aren't code that you'll see outside of the old code you saw 😄 . Also the "good" example is different than the pattern we have been using everywhere so it'll probably cause confusion.

I think we should probably be more intention with making the change to use property arrow functions like you have here. Otherwise there'll be lots of confusion I'm afraid

Comment (jiangyue-eb) [reply es6/README.md]: @benmvp That makes sense. Is there a place here to suggest "better" rather than using `good` and `bad`? e.g. If need to use event handlers, better to use fat arrow syntax with the reasons listed above. I feel that's the real intention from this PR.

Comment (jiangyue-eb) [reply es6/README.md]: Or I can just replace my addition with some suggestions and put examples there.

Comment (kaylieEB) [reply es6/README.md]: We can rephrase this to `When defining methods to be attached to event listeners, we recommend using the fat arrow syntax within classes to bind the current context without returning a new function each time.`

I actually prefer us moving towards this approach over the current `.bind` but don't expect anyone to refactor their old code. 

Also nit, we should use `PureComponent` instead of `Component` for consistency.

Comment (jiangyue-eb) [reply es6/README.md]: @kaylieEB 👍 I will update my PR.

Comment (jiangyue-eb) [reply es6/README.md]: Updated in https://github.com/eventbrite/javascript/pull/55/commits/45bd051c71a099f70b0808b3667e7b05c33388ca

File: es6/README.md [hunk 4]
```diff
@@ -704,6 +704,42 @@ Button.propTypes = {
 
 Both declarative property syntaxes are not a part of the ES2015 specification and are a in the midst of the ECMAScript proposal approval process. Currently they are sitting in Stage 3. For details, check out [ECMAScript Class Fields and Static Properties](https://github.com/jeffmo/es-class-fields-and-static-properties).
 
+When defining methods to be attached to event listeners, we recommend using the fat arrow syntax within classes to bind the current context without returning a new function each time.
+
+```js
+
+// good: bind the current context using fat arrow syntax
+class myComponent extends React.PureComponent {
+    onResize = () => {
```
Comment (kaylieEB) [inline es6/README.md]: can you rename to `_handleResize` to be consistent with our other conventions?

File: es6/README.md [hunk 5]
```diff
@@ -704,6 +704,42 @@ Button.propTypes = {
...
+
+// bad: use `::func` to bind events need to be removed later
+// the `resize` handler won't be removed as `::onResize` will return a new function
+class myComponent extends React.PureComponent {
+    onResize() {
+        //handle resize window.
+    }
+
+    componentDidMount() {
+        window.addEventListener('resize', ::this.onResize);
```
Comment (kaylieEB) [inline es6/README.md]: I think @BenAtEventbrite meant we should keep the other `bad` example and get rid of this one since only seat designer uses the bind operator and other places uses `.bind` with `this.handleResize = this._handleResize.bind(this)`

Comment (jiangyue-eb) [reply es6/README.md]: hmm, ok, I mean if we do `this.handleResize = this._handleResize.bind(this)` in the constructor then later reference to `this.handleResize` then it is not "too bad" compared to use`::` since it is not introducing bugs. But I feel it's fine if `seat_designer` is the only place having this syntax, and I've started to refactor some of them, hopefully later all will be gone.
Do we want to keep the `::` example just in case people later want to use it for some reasons? 😆 

I will restore the example of `this.handleResize = this._handleResize.bind(this)`

Comment (jiangyue-eb) [reply es6/README.md]: updated
